external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2937,12 +2940,82 @@ fn module_call_impl(
     (call, return_)
 }
 
+/// If the constant that we are referencing has an `error` attribute, we need to change the type of
+/// the constant to a u64 since this will be compiled into a u64 error code.
+fn annotated_error_const(context: &mut Context, e: &mut T::Exp) {
+    if let sp!(
+        const_loc,
+        T::UnannotatedExp_::Constant(module_ident, constant_name)
+    ) = &mut e.exp
+    {
+        let ConstantInfo {
+            attributes,
+            defined_loc,
+            signature,
+        } = context.constant_info(module_ident, constant_name);
+        let has_error_annotation =
+            attributes.contains_key_(&known_attributes::ErrorAttribute.into());
+        let u64_type = matches!(
+            signature.value.builtin_name(),
+            Some(sp!(_, N::BuiltinTypeName_::U64))
+        );
+        let ty_loc = e.ty.loc;
+        let const_def_loc = *defined_loc;
+
+        if has_error_annotation
+            && context.env.check_feature(
+                context.current_package(),
+                FeatureGate::CleverAssertions,
+                *const_loc,
+            )
+        {
+            e.ty = N::Type_::builtin(ty_loc, sp(ty_loc, N::BuiltinTypeName_::U64), vec![]);
```

Comment:
```suggestion
        {
            context.env.check_feature(
                context.current_package(),
                FeatureGate::CleverAssertions,
                *const_loc,
            );
         }
         if supports_feature {
            e.ty = N::Type_::builtin(ty_loc, sp(ty_loc, N::BuiltinTypeName_::U64), vec![]); 
         }
```

Reply:
I think checking the feature on `has_error_annotation` feels a bit more resilient. They are trying to use the feature, but cant. Everything else at this point should assume its enabled

Reply:
I also think we should always swap it to u64 with the feature enabled, it will lead to better error messages since you don't get double hit with the error for no `#[error]` and then again for non-u64